### PR TITLE
fix #23, add support for PS256, some simple tests

### LINKED ISSUE #23: Don't hard code signature algorithm

### PR DESCRIPTION
**Pull request by mozkeeler**

### CODE REVIEW

File: src/cose/test_cose.rs [hunk 1]
```diff
@@ -206,3 +206,63 @@ fn test_cose_sign_verify_tampered_signature() {
     assert!(verify_result.is_err());
     assert_eq!(verify_result, Err(CoseError::VerificationFailed));
 }
+
+#[test]
+fn test_cose_sign_verify_rsa() {
```
Comment (franziskuskiefer) [inline src/cose/test_cose.rs]: I feel like these all the tests in this file duplicate a lot of code. We should probably make them share some of the code.

File: src/cose/nss.rs [hunk 2]
```diff
@@ -277,14 +278,19 @@ pub fn sign(
...
     let hash_item = SECItem::maybe_new(hash_buf.as_slice())?;
     let signature_len = unsafe { PK11_SignatureLen(key) };
-    let mut signature: Vec<u8> = Vec::with_capacity(signature_len);
-    // XXX: rewrite without unsafe.
-    unsafe {
-        signature.set_len(signature_len);
-    }
+    // We indirectly pass this to PK11_SignWithMechanism and it gets modified,
+    // so this has to be mut.
+#[allow(unused_mut)]
```
Comment (franziskuskiefer) [inline src/cose/nss.rs]: Hm, I expected this to work. But I still get the warning. It works when I put it on the function (but that's probably not what we want).

File: src/cose/test_cose.rs (hunk 1 above)
Comment (mozkeeler) [reply src/cose/test_cose.rs]: Ok - I filed #36 for this.

File: src/cose/nss.rs (hunk 2 above)
Comment (mozkeeler) [reply src/cose/nss.rs]: Yeah, it turns out this wasn't the right approach anyway. I'll push an update shortly.

**Comment by mozkeeler** [comment]
@franziskuskiefer let me know what you think about the interdiff. If that looks alright, feel free to merge so we can move forward on #35.

File: src/cose/nss.rs [hunk 3]
```diff
@@ -34,6 +34,45 @@ impl SECItem {
...
+    /// The input vec is not expected to have any actual contents, and in any case is cleared.
+    fn maybe_from_vec(vec: &'a mut Vec<u8>) -> Result<SECItemMut<'a>, NSSError> {
+        if vec.len() > u32::max_value() as usize {
+            return Err(NSSError::InputTooLarge);
+        }
+        vec.clear();
+        Ok(SECItemMut {
+            typ: SI_BUFFER,
+            data: vec.as_mut_ptr(),
+            len: vec.capacity() as u32,
```
Comment (franziskuskiefer) [inline src/cose/nss.rs]: `capacity` returns the number of elements the vector can hold without reallocating. I think you want to use `len` here.

File: src/cose/nss.rs [hunk 4]
```diff
@@ -34,6 +34,45 @@ impl SECItem {
...
+    /// SECItemMut that points to the vec and has the same capacity.
+    /// The input vec is not expected to have any actual contents, and in any case is cleared.
+    fn maybe_from_vec(vec: &'a mut Vec<u8>) -> Result<SECItemMut<'a>, NSSError> {
+        if vec.len() > u32::max_value() as usize {
+            return Err(NSSError::InputTooLarge);
+        }
+        vec.clear();
+        Ok(SECItemMut {
+            typ: SI_BUFFER,
+            data: vec.as_mut_ptr(),
```
Comment (franziskuskiefer) [inline src/cose/nss.rs]: This way the vector has to stick around for the `SECItemMut` not to become garbage. Is this intended?

File: src/cose/nss.rs (hunk 3 above)
Comment (franziskuskiefer) [reply src/cose/nss.rs]: Ah, after reading the rest I understand how this works. But I think it's a little confusing. `maybe_from_vec` only works for vectors that have a capacity set to the size we need here. Maybe we should call this function differently to make clear what it's doing?

File: src/cose/nss.rs (hunk 4 above)
Comment (mozkeeler) [reply src/cose/nss.rs]: Yeah - the idea is the PhantomData and the lifetime argument takes care of this.

**Pull request merged**